misk-hibernate-testing/src/main/kotlin/misk/jdbc/DatabaseAllocator.kt

```diff
@@ -0,0 +1,213 @@
+package misk.jdbc
+
+import misk.hibernate.Session
+import misk.hibernate.Transacter
+import misk.hibernate.shards
+import misk.hibernate.transaction
+import java.time.Clock
+import java.time.Duration
+import javax.inject.Inject
+import javax.inject.Provider
+import javax.inject.Singleton
+import javax.persistence.PersistenceException
+import kotlin.reflect.KClass
+
+/**
+ * This is used to allocate a new database named like `[hint]__[now]__[latestVersion]` using the
+ * base name specified by a [DataSourceConfig].
+ *
+ * [DatabaseAllocator]s are to be used _only_ in tests, where we might want to have create and use
+ * a new database between tests to avoid parallelism issues.
+ *
+ */
+@Singleton
+class DatabaseAllocator(
+  val qualifier: KClass<out Annotation>,
+  val backend: Backend,
+  val clock: Clock,
+  config: DataSourceConfig,
+  retention: Duration = Duration.ofDays(2)
+) {
+  init {
+    backend.type = config.type
+  }
+
+  private lateinit var currentDatabase: String
+  private val hint: String = config.database!!
+  private var latestVersion: String = "1"
+
+  private val retentionMillis: Long = retention.toMillis()
+  private val now: Long
+    get() = clock.instant().toEpochMilli()
+
+  private fun nextVersion(): String {
+    latestVersion = "${latestVersion.toInt() + 1}"
+    return latestVersion
+  }
+
+  /** Returns a list of all databases that were created by a database allocator. */
+  fun existingDatabases(): List<String> {
+    return backend.showDatabases()
+        .filter { it.matches(Regex(".*__[0-9]+__[0-9]+")) }
+  }
+
+  /** Returns the name of the database most recently allocated by this. */
+  fun currentDatabase(): String = currentDatabase
+
+  /**
+   * Given a database name as a hint, allocates a database called `[hint]__[now]__[latestVersion]`
+   * and returns the name of the allocated database.
+   */
+  fun allocateDatabase(): String {
+    val databases = existingDatabases()
+    latestVersion = databases
+        .filter { it.startsWith(hint) }
+        .max()
+        ?.split("__")
+        ?.last() ?: latestVersion
+
+    var allocated = false
+    var databaseName = "${hint}__${now}__$latestVersion"
+
+    // Keep trying to create a database until we have found an unused name.
+    while (!allocated) {
+      try {
+        databaseName = "${hint}__${now}__$latestVersion"
+        backend.createDatabase(databaseName)
+        allocated = true
+      } catch (e: PersistenceException) {
+        nextVersion()
+      }
+    }
+
+    currentDatabase = databaseName
+    return databaseName
+  }
+
+  /**
+   * Drops all databases that were created by an allocator which are older than the retention
+   * duration of this allocator.
+   */
+  @Throws(PersistenceException::class)
+  fun dropOldDatabases() {
```

Comment:
pruneOldDatabases ?

We like the word `prune` when removing things because they’ve been expired from a cache

Reply:
Updated!